### DISCUSSION TITLE
아이템 67. 최적화는 신중히 하라

### DISCUSSION THREAD

**Discussion by chikeem90** [Study-2-Effective-Java/Effective-Java#167] [3. 과제]
> 컴퓨터에서의 대부분의 죄악은 섣부른 효율성이라는 이름으로 저질러진다. 
> William A. Wulf

> 우리는 작은 효율성에 대해서는 잊어버려야 한다. 섣부른 최적화는 만악의 근원이다. 
> Donald E. Knuth

> 최적화라는 명목으로 우리는 다음 두 가지 규칙을 따른다. 
> 규칙 1. 최적화 하지 마라
> 규칙 2. (오직 전문가만) 아직 최적화 하지 마라, 즉 완벽하게 명백한 최적화되지 않은 해결책이 있지 않는 이상 최적화 하지마라.
> M. A. Jackson

### (성능에만 집중하기 보다는) 좋은 프로그램을 설계하라
- 섣부른 최적화는 오히려 독이 될 수 있음.
- 성능 같은 구현 상의 문제는 나중에 최적화해서 해결할 수 있지만, 아키텍처의 결함은 시스템 전체를 다시 작성하지 않고는 해결하기 불가능할 수 있음.
- 좋은 프로그램이지만 원하는 성능이 나오지 않는다면, 그 아키텍처 자체가 최적화할 수 있는 길을 안내해줄 것.

```java
public abstract class Component implements ImageObserver, MenuContainer,
                                           Serializable
{
    public Dimension getSize() {
        return size();
    }

    @Deprecated
    public Dimension size() {
        return new Dimension(width, height); // 방어적 복사
    }
}
```

- 애초에 `Dimension`을 가변으로 설계했기 때문에 방어적 복사를 고려해야하고 이는 성능상 영향을 줄 수 있음.

### 성능 최적화를 하게될 경우 최적화 전후 성능을 측정하라
- 최적화를 했음에도 불구하고 성능이 생각보다 좋아지지 않는 경우가 있으며, 심지어 성능을 더 나빠지게 할 수도 있음.
- 프로파일링 도구를 활용할 것. (ex. jmh)

### 핵심 정리
- 성능보다는 견고한 구조의 프로그램을 작성하자
- 견고한 구조의 프로그램을 작성하다보면 성능은 자연스레 따라온다
- 시스템을 구현했다면 성능을 측정해보자. 충분히 빠르다면 최적화를 하지 않아도 괜찮다
- 만약 최적화를 한다면 문제의 원인이 되는 부분을 파악한 후 최적화를 하자

**Comment by Irisation23** [comment]
```
규칙 1. 최적화 하지 마라 
```

앞으로 내 꿈은 너야 박 Jackson

**Comment by corock** [comment]
이번 장의 주제는 처음 봤을 때 최적화하지 말라는 건가? 로 보게 되어 좀 의외였는데요.
하지만 내용 중에 "성능보다 좋은 구조를 고안하라" 라는 문장이 깊은 울림을 가져다주었습니다.
평소 코딩할 때 어떤 것에 중점을 두고 해야 하는지 되짚어 보는 아이템이었네요. 잘 봤습니다~

**Comment by YuDeokRin** [comment]
프로파일링 도구 jmh 툴을 알게되었네요. (이런것도있구나..)
